### DISCUSSION TITLE
Fitur Multi-Device Dirilis - v1.0-RC14

### DISCUSSION THREAD

**Discussion by rioastamal** [kirimwa-id/forum#120] [Pengumuman]
Halo semua,

Bertepatan dengan ulang tahun API KirimWA.id yang ke-1, saya dengan sangat senang mengumumkan bahwa fitur **Multi-Device** telah hadir di API KirimWA.id **v1.0-RC14**. 

Saya ucapkan banyak terima kasih kepada semua pengguna API KirimWA.id yang telah sabar menunggu fitur ini sejak awal tahun 2022. Saya menerima cukup banyak pertanyaan tentang fitur ini baik via email dan atau pun lewat thread diskusi #42. Hal ini tidak lepas dari kebijakan WhatsApp yang melakukan _force update_ versi yang belum MD ke MD sehingga tidak bisa lagi menggunakan API KirimWA.id

## Cara Konfigurasi

Tidak ada perubahan API atau konfigurasi pada rilis ini. Satu hal yang perlu anda lakukan hanyalah melakukan scan ulang QR code agar pairing ke versi multi-device terjadi.

```sh
curl -s 'https://api.kirimwa.id/v1/qr?device_id=DEVICE_ID'
```

```json
{
  "qr_code": "1@U2tS5Q1elzj6Y7IAKDurwvja47SQz8bvW24fb43r3n+gPC4PN1iSNlGLizlfGsrHSD/M6ym6/aYiYw==,xXlcsqVquva7/1c2g8wAZkWdnk2el5tHWh7MWUW2UTc=,v+P3exsbB1W62wX3Vn4dcC==",
  "image_url": "https://api.kirimwa.id/v1/qr/show?qrcode=1%40U2tS5Q1elzj6Y7IAKDurwvja47SQz8bvW24fb43r3n%2BgPC4PN1iSNlGLizlfGsrHSD%2FM6ym6%2FaYiYw%3D%3D%2CxXlcsqVquva7%2F1c2g8wAZkWdnk2el5tHWh7MWUW2UTc%3D%2Cv%2BP3exsbB1W62wX3Vn4dcC%3D%3D&device_id=YOUR_DEVICE_ID"
}
```

### Catatan

- Ini adalah versi pertama dari multi device yang saya rilis, jadi mungkin ada beberapa bug yang muncul. Silahkan laporkan lewat thread diskusi.
- Sejak versi ini v1.0-RC14 maka API KirimWA.id hanya mendukung versi WhatsApp yang memiliki fitur Multi-Device.

Saya harap dengan fitur ini akan mempermudah teman-teman developer dalam mengembangkan aplikasi yang mengandalkan API KirimWA.id.

Belum memiliki Token API KirimWA.id? Daftar gratis di https://developer.kirimwa.id/

Salam,
Rio Astamal

**Comment by ekoyuliawan** [comment]
Halo mas Rio, terima kasih banyak atas rilis API WA versi multi-device yang sudah lama kami tunggu-tunggu. Segera mencobanya! 🥇

**Comment by melodiazz** [comment]
Thanks a lot Pak Rio..

**Comment by sanusi86** [comment]
Mantaps mas Rio
Saya sudah coba scan QR dan berhasil masuk, tapi status masih disconnect, mohon pencerahannya

**Reply by melodiazz** [reply]
saya juga pertama kali coba seperti itu & tidak bisa kirim message (meskipun status di linked device masih aktif), kemudian coba rescan qr code status masih disconnected tapi coba kirim message sudah lancar

**Reply by sanusi86** [reply]
baik, terima kasih
saya coba kirim message

**Reply by rioastamal** [reply]
Terima kasih laporan bug-nya. Meskipun disconnected coba saja untuk mengirim pesan, harusnya pesan masih bisa terkirim selama status di-linked device masih active.

**Reply by sanusi86** [reply]
pesan sukses terkirim walaupun status disconnected

**Reply by sanusi86** [reply]
send message OK, Webhook OK. 
mantaps, terima kasih mas Rio

**Reply by rioastamal** [reply]
Fixed pada v1.0-RC15. Sekarang harusnya status sudah sesuai. #20

**Reply by lutfiihsan** [reply]
@rioastamal fitur reconnect device tidak bekerja mas. Saya coba reconnect device. Setelah saya coba send message, saya mendapat pesan error "no authenticate".

**Reply by lutfiihsan** [reply]
@rioastamal mas ketika saya connect device baru, ada batasan waktu nya gk mas, sehingga kita harus reconnect?

**Reply by rioastamal** [reply]
@lutfiihsan jika device tidak terkoneksi dengan internet maka menurut teori dari WhatsApp, 14 hari baru akan putus. Coba lihat status devicesnya lewat API `/devices/DEVICE_ID` atau pasang webhook. Webhook otomatis akan dikirim jika ada perubahan status perangkat lihat https://developer.kirimwa.id/api/v1/#webhook-device-status

**Reply by lutfiihsan** [reply]
@rioastamal apakah tokennya tidak disimpan di db atau session mas?

**Reply by rioastamal** [reply]
Semua disimpan didatabase baik API Token maupun state dari koneksi

**Reply by lutfiihsan** [reply]
@rioastamal mas untuk dokumentasi send image, file blm ada y mas?

**Reply by rioastamal** [reply]
Maksudnya send file dari local disk bukan URL?

**Reply by lutfiihsan** [reply]
Iya dr local mas

**Reply by shahib-mirbath** [reply]
@lutfiihsan
bisa baca dokumen di link berikut: https://developer.kirimwa.id/api/v1/#post-messages
kirim file bisa. tp harus harus berupa link yg dapat diakses server api.kirimwa.id

**Comment by shahib-mirbath** [comment]
respon yg awalnya qr_code jadi qr saja ya mas @rioastamal ?

**Reply by kafaa-dev** [reply]
iya mas

**Reply by rioastamal** [reply]
Terima kasih laporannya.

**Reply by rioastamal** [reply]
Fixed pada v1.0-RC15 #20

**Comment by shahib-mirbath** [comment]
wa yg dikirim melalui device atau web masuk webhook incoming_message.

**Reply by rioastamal** [reply]
Ini maksudnya gimana mas @shahib-mirbath?

**Comment by kafaa-dev** [comment]
#121
Mohon pencerahannya mas @rioastamal 🙏

**Comment by azmirizkifar20** [comment]
Alhamdulillah.. terimakasih banyak mas @rioastamal

**Comment by zonegoh** [comment]
terima kasih mas @rioastamal

**Comment by perogeremmer** [comment]
GOKIL! Thanks @rioastamal! 👍🏻

**Comment by shahib-mirbath** [comment]
Tidak mengirim pesan saat device mati padahal status di get device itu connected.

**Reply by rioastamal** [reply]
Saya tidak menjumpai masalah itu di perangkat saya. Saya coba matikan koneksi internet (airplane mode) selama beberapa jam dan tetap bisa mengirim.

Coba logout saja dari device kemudian scan qr ulang.

**Reply by shahib-mirbath** [reply]
Baik, mas

**Comment by shahib-mirbath** [comment]
POST Message selain text, jika **caption** tidak diikutkan, maka akan terisi waktu server. Seperti gambar berikut:
![image](https://user-images.githubusercontent.com/46096681/180611385-88334cd4-ddf3-4084-88bf-a6f638574cd2.png)

**Reply by rioastamal** [reply]
Terima kasih laporannya, akan coba saya cek.

**Comment by masbagusstwn** [comment]
Halo mas @rioastamal ,

Untuk responses dari get-incoming-messages apakah memang berbeda dengan di docs?

di saya saat coba di get posisinya agak tidak beraturan tidak seperti di contoh sini https://developer.kirimwa.id/api/v1/#get-incoming-messages

output nya seperti ini :

{
  "data": [
    {
      "location": {
        "name": null,
        "address": null,
        "latitude": 0,
        "longitude": 0
      },
      "created_at": "2022-07-24T13:54:26.606Z",
      "is_group_message": false,
      "meta": {
        "seconds": 0,
        "size": 10,
        "mime_type": "plain/text",
        "dimension": {
          "width": 0,
          "height": 0
        },
        "file_name": null
      },
      "contact": {
        "name": null,
        "vcard": null
      },
      "timestamp": 1658670866,
      "message_type": "text",
      "device_id": "demos",
      "text": "yg ada aja",
      "group_id": null,
      "sender": "62857762xxx",
      "caption": null,
      "from_me": true,
      "id": "3EB027C0CC4xxx"
    },

**Reply by rioastamal** [reply]
@masbagusstwn , bedanya dimana?

Untuk posting snippet kode apit dengan tiga backtick agar mudah dibaca.

```json
{
  "contoh": "biar mudah dibaca"
}
```

Cek disini untuk lengkapnya https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

**Reply by masbagusstwn** [reply]
maksudnya beda struktur object / field nya mas. kalo di docs nya Mas Rio posisi id diatas lalu selanjutnya data datanya kebawah.

tapi pas saya coba get messages hasilnya id malah di bawah seperti berikut

```
{
  "data": [
      {
        "location": {
            "name": null,
            "address": null,
            "latitude": 0,
            "longitude": 0
      },
      "created_at": "2022-07-24T13:54:26.606Z",
      "is_group_message": false,
      "meta": {
           "seconds": 0,
           "size": 10,
           "mime_type": "plain/text",
           "dimension": {
                "width": 0,
                "height": 0
           },
           "file_name": null
      },
      "contact": {
           "name": null,
           "vcard": null
      },
      "timestamp": 1658670866,
      "message_type": "text",
      "device_id": "demos",
      "text": "yg ada aja",
      "group_id": null,
      "sender": "62857762xxx",
      "caption": null,
      "from_me": true,
      "id": "3EB027C0CC4xxx"
  },
```

lebih ke posisi aja si mas beda, secara content atau isi tetap sama setelah saya teliti

**Comment by teknisilaptopbali** [comment]
> Halo semua,
> 
> Bertepatan dengan ulang tahun API KirimWA.id yang ke-1, saya dengan sangat senang mengumumkan bahwa fitur **Multi-Device** telah hadir di API KirimWA.id **v1.0-RC14**.
> 
> Saya ucapkan banyak terima kasih kepada semua pengguna API KirimWA.id yang telah sabar menunggu fitur ini sejak awal tahun 2022. Saya menerima cukup banyak pertanyaan tentang fitur ini baik via email dan atau pun lewat thread diskusi #42. Hal ini tidak lepas dari kebijakan WhatsApp yang melakukan _force update_ versi yang belum MD ke MD sehingga tidak bisa lagi menggunakan API KirimWA.id
> 
> ## Cara Konfigurasi
> Tidak ada perubahan API atau konfigurasi pada rilis ini. Satu hal yang perlu anda lakukan hanyalah melakukan scan ulang QR code agar pairing ke versi multi-device terjadi.
> 
> ```shell
> curl -s 'https://api.kirimwa.id/v1/qr?device_id=DEVICE_ID'
> ```
> 
> ```json
> {
>   "qr_code": "1@U2tS5Q1elzj6Y7IAKDurwvja47SQz8bvW24fb43r3n+gPC4PN1iSNlGLizlfGsrHSD/M6ym6/aYiYw==,xXlcsqVquva7/1c2g8wAZkWdnk2el5tHWh7MWUW2UTc=,v+P3exsbB1W62wX3Vn4dcC==",
>   "image_url": "https://api.kirimwa.id/v1/qr/show?qrcode=1%40U2tS5Q1elzj6Y7IAKDurwvja47SQz8bvW24fb43r3n%2BgPC4PN1iSNlGLizlfGsrHSD%2FM6ym6%2FaYiYw%3D%3D%2CxXlcsqVquva7%2F1c2g8wAZkWdnk2el5tHWh7MWUW2UTc%3D%2Cv%2BP3exsbB1W62wX3Vn4dcC%3D%3D&device_id=YOUR_DEVICE_ID"
> }
> ```
> 
> ### Catatan
> * Ini adalah versi pertama dari multi device yang saya rilis, jadi mungkin ada beberapa bug yang muncul. Silahkan laporkan lewat thread diskusi.
> * Sejak versi ini v1.0-RC14 maka API KirimWA.id hanya mendukung versi WhatsApp yang memiliki fitur Multi-Device.
> 
> Saya harap dengan fitur ini akan mempermudah teman-teman developer dalam mengembangkan aplikasi yang mengandalkan API KirimWA.id.
> 
> Belum memiliki Token API KirimWA.id? Daftar gratis di https://developer.kirimwa.id/
> 
> Salam, Rio Astamal
![WhatsApp Image 2022-07-26 at 9 23 20 PM](https://user-images.githubusercontent.com/110038645/181016533-1ee1d381-8cca-4dbb-84a5-276193f671f6.jpeg)
Hi mas saat mau scan QRcode invalid QRcode ada solusi kah?

**Reply by rioastamal** [reply]
Pastikan saat melakukan scan tidak lebih dari 10-15 detik dari pemanggilan `/qr`. Sertakan juga versi OS dari HP dan WhatsApp yang digunakan.

**Comment by HanustaviraGA** [comment]
![WhatsApp Image 2022-09-14 at 21 22 09](https://user-images.githubusercontent.com/68433367/190181102-9a5260d2-cc01-4c9c-b75c-a71724b77a49.jpeg)

Mas @rioastamal mohon bantuannya ini saya sudah coba scan tapi tidak terdeteksi dan tidak beraksi apa2, kira2 kenapa ya ? Terima kasih
